File: Code/Erlang/problem017/euler017.erl
-module(euler017).
-export([
    euler/1
    ,number_to_word/1
    ,number_to_letter_count/1
  ]
).

euler(N) when N > 0, N =< 1000, is_integer(N) -> 
  length( 
    lists:flatmap( 
      fun(X) -> number_to_word(X) end,
      lists:seq(1,N)
    )
  ).

number_to_letter_count(N) when N > 0, N =< 1000, is_integer(N) ->
  length( number_to_word(N) ).

number_to_word(N) when N > 0, N =< 1000, is_integer(N) ->
  if 
    N < 10 -> one_to_nine_to_word(N);
    N >= 10, N <  20 -> teens_to_word(N);
    N >= 20, N =< 99 -> 
      Remdr  = (N rem 10),
      Decade = N - Remdr,
      decade_to_word(Decade) ++ one_to_nine_to_word(Remdr);
    N >= 100, N < 1000 -> 
      Basic = N rem 100,
      Century = (N - Basic) div 100,
      if
        Basic == 0 -> Ender = "";
        Basic /= 0 -> Ender = "and" ++ number_to_word(Basic)
      end,
      one_to_nine_to_word(Century) ++ "hundred" ++ Ender;
    N == 1000 -> "onethousand"
  end.

one_to_nine_to_word(N) when N >= 0, N =< 9, is_integer(N) ->
  case N of
    0 -> "";
    1 -> "one";
    2 -> "two";
    3 -> "three";
    4 -> "four";
    5 -> "five";
    6 -> "six";
    7 -> "seven";
    8 -> "eight";
    9 -> "nine"
  end.

teens_to_word(N) when N >= 10, N =< 19, is_integer(N) ->
  case N of
    10 -> "ten";
    11 -> "eleven";
    12 -> "twelve";
    13 -> "thirteen";
    14 -> "fourteen";
    15 -> "fifteen";
    16 -> "sixteen";
    17 -> "seventeen";
    18 -> "eighteen";
    19 -> "nineteen"
  end.

decade_to_word(N) when N >= 20, N =< 90, is_integer(N) ->
  case N of
    20 -> "twenty";
    30 -> "thirty";
    40 -> "forty";
    50 -> "fifty";
    60 -> "sixty";
    70 -> "seventy";
    80 -> "eighty";
    90 -> "ninety"
  end.

File: Code/Python/problem017/euler017.py
#!/usr/bin/env python

one_to_nine = {0: "", 1: "one", 2: "two", 3: "three", 4: "four", 5: "five",
               6: "six", 7: "seven", 8: "eight", 9: "nine"}

teens = {10: "ten", 11: "eleven", 12: "twelve", 13: "thirteen",
         14: "fourteen", 15: "fifteen", 16: "sixteen", 17: "seventeen",
         18: "eighteen", 19: "nineteen"}

decades = {20: "twenty", 30: "thirty", 40: "forty", 50: "fifty", 60: "sixty",
           70: "seventy", 80: "eighty", 90: "ninety"}


def number_to_word(num):
    if num > 0 and num < 10:
        return one_to_nine[num]
    elif num >= 10 and num < 20:
        return teens[num]
    elif num >= 20 and num < 100:
        rem = num % 10
        num = num - rem
        return decades[num] + one_to_nine[rem]
    elif num >= 100 and num < 1000:
        simple = num % 100
        century = int((num - simple) / 100)
        if simple == 0:
            ender = ""
        else:
            ender = "and" + number_to_word(simple)
        return one_to_nine[century] + "hundred" + ender
    if num == 1000:
        return "onethousand"
    return ""


def euler():
    lettered = ""
    for i in range(1, 1001):
        lettered += number_to_word(i)
    print len(lettered)


if __name__ == '__main__':
    euler()
